ambry-cloud/src/main/java/com/github/ambry/cloud/azure/AzureCloudDestinationSync.java

```diff
@@ -680,9 +687,107 @@ public FindResult findEntriesSince(String partitionPath, FindToken findToken, lo
+        long expirationTime = Long.parseLong(metadata.get(CloudBlobMetadata.FIELD_EXPIRATION_TIME));
+        eraseBlob = (expirationTime + gracePeriod) < now;
+        eraseReason = String.format("%s: (%s + %s) < %s", CloudBlobMetadata.FIELD_EXPIRATION_TIME, expirationTime, gracePeriod, now);
+      }
+
+      if (eraseBlob) {
+        if (cloudConfig.cloudCompactionDryRunEnabled) {
+          logger.trace("[DRY-RUN] Can erase blob {} from Azure blob storage because {}", blobItem.getName(), eraseReason);
+        } else {
+          Timer.Context storageTimer = azureMetrics.blobCompactionLatency.time();
+          logger.trace("Erasing blob {} from Azure blob storage because {}", blobItem.getName(), eraseReason);
+          blobContainerClient.getBlobClient(blobItem.getName()).delete();
+          storageTimer.stop();
+          vcrMetrics.blobCompactionRate.mark();
+        }
+        numBlobsPurged += 1;
+      } else {
+        logger.trace("Cannot erase blob {} from Azure blob storage because condition not met: {}", blobItem.getName(), eraseReason);
+      }
+
+    }
+    return numBlobsPurged;
+  }
+
+  /**
+   * Compacts a partition
+   * @param partitionPath the path of the partitions to compact.
+   * @return the number of blobs erased from cloud
+   * @throws CloudStorageException
+   */
   @Override
   public int compactPartition(String partitionPath) throws CloudStorageException {
-    return 0;
+    /*
+      BlobContainerStrategy must be PARTITION, otherwise compaction will not work because it cannot find the container in ABS.
+      For BlobContainerStrategy to be CONTAINER, we need a blobId to extract accountId and containerId and we don't have that here.
+     */
+    if (AzureBlobLayoutStrategy.BlobContainerStrategy.get(azureCloudConfig.azureBlobContainerStrategy) !=
+        AzureBlobLayoutStrategy.BlobContainerStrategy.PARTITION) {
+      logger.info("Unable to compact because BlobContainerStrategy is {} when it must be {}", azureCloudConfig.azureBlobContainerStrategy,
+          AzureBlobLayoutStrategy.BlobContainerStrategy.PARTITION);
+      return 0;
+    }
+    String containerName = azureBlobLayoutStrategy.getClusterAwareAzureContainerName(partitionPath);
+    BlobContainerClient blobContainerClient = createOrGetBlobStore(containerName);
+    ListBlobsOptions listBlobsOptions = new ListBlobsOptions().setDetails(new BlobListDetails()
+        .setRetrieveMetadata(true)).setMaxResultsPerPage(azureCloudConfig.azureBlobStorageMaxResultsPerPage);
+    String continuationToken = null;
+    int numBlobsPurged = 0, totalNumBlobs = 0;
+    Timer.Context storageTimer = azureMetrics.partitionCompactionLatency.time();
+    try {
+      logger.info("Initiating compaction of partition {} in Azure blob storage", containerName);
+      for (PagedResponse<BlobItem> blobItemPagedResponse :
+          blobContainerClient.listBlobs(listBlobsOptions, null).iterableByPage(continuationToken)) {
```

Comment:
Just a question. I was just looking at v1 code. It looks like we are checkpointing compaction progress(I think it may be have easy there since we can provide the next start time as query to cosmos). 

Would we have a way to save compaction progress here too? If not, could we hit a case where compaction progresses slowly since we have to restart from beginning every time a vcr node restarts?

Reply:
Contrary to popular belief, compaction is not slow. Disk-compaction cannot be compared to cloud-compaction and prev impl of cloud-compaction has lot of unnecessary metadata accesses. I made some measurements early on for  the time it takes to get metadata of a single video-cluster partition and here are the results. We can look into checkpointing state if the perf is unacceptable after deploying in prod.

2023/07/15 03:47:18.598 INFO [RecoveryNetworkClient] [RecoveryThread-Intra-0-prod-lva1] [ambry-server] [] |snkt| ABS | numPages = 2500, numBlobs = 250000, resultFetchTime = 88709 ms @ 100 items/page

2023/07/15 03:31:53.285 INFO [RecoveryNetworkClient] [RecoveryThread-Intra-0-prod-lva1] [ambry-server] [] |snkt| ABS | numPages = 57, numBlobs = 252431, resultFetchTime = 103392 ms @5000 items/page